AirQuality binding

kubawolanin opened the pull request:
So here's my attempt to utilize World Air Quality Index data in our home automation systems :)
It's my first binding, so I'm sure there are some things to improve.
i hope you'll like it!

I'm attaching compiled JAR file to this post. It will be updated accordingly.
Please make sure you change extension from `*.zip` to `*.jar` before you drop it to the `\addons\` folder.

Updated JAR with latest (**04.02.2017**) code review follow-up:
[org.openhab.binding.airquality-2.1.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/752206/org.openhab.binding.airquality-2.1.0-SNAPSHOT.zip)

Big shoutout to @lolodomo whose Wunderground binding helped me a lot with understanding all the nuances in building one :) Thank you!

Signed-off-by: Kuba Wolanin <hi@kubawolanin.com>
Also-by: Łukasz Dywicki <luke@code-house.org>

hakan42 commented on addons/binding/org.openhab.binding.airquality/.gitignore:
```diff
@@ -0,0 +1,5 @@
+.classpath
```
Please remove binding-specific gitignore file from the PR. We should depend on the overall .gitignore file in the root of the project ( @kaikreuzer : I know we are using eclipse, but could we add the .idea and .*.iml for IntelliJ anyway?)

hakan42 commented on addons/binding/org.openhab.binding.airquality/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,164 @@
+        </state>
+    </channel-type>
+
+</thing:thing-descriptions>
```
Could you add a line feed at the end of the file? Makes things nicer to read on the shell.

kubawolanin replied:
Done! Thanks

hakan42 commented on addons/binding/org.openhab.binding.airquality/README.md:
````diff
@@ -0,0 +1,118 @@
+airquality.things:
+
+```
+airquality:aqi:Krakow "AirQuality in Krakow" [ apikey="XXXXXXXXXXXX", location="50.06465,19.94498", refresh=60 ]
````
I just discovered that you can add a Location in the thing description 😄 

Could you change the description to "AirQualitry" @ "Krakow" so we have examples demostrating this syntax?

rtvb replied:
The thing uses 'Krakow' where the items use 'krk'

rtvb replied:
I would use airquality:aqi:home for easy copy and pasting by users.

hakan42 commented on addons/binding/org.openhab.binding.airquality/README.md:
````diff
@@ -0,0 +1,118 @@
+
+```
+sitemap demo label="Air Quality"
+{
````
opening brace to the upper line please (so all files look similiar)

hakan42 commented on addons/binding/org.openhab.binding.airquality/README.md:
````diff
@@ -0,0 +1,118 @@
+airquality.sitemap:
+
+```
+sitemap demo label="Air Quality"
````
Please call the sitemap "airquality" so it matches the filename. The Android app (or maybe Rotini or 3House?) was pretty picky about it.

ThomDietrich replied:
I believe this to be a general requirement. Please anyone correct me if that's not strictly correct.

rtvb replied:
@ThomDietrich @hakan42 you are correct

hakan42 commented on addons/binding/org.openhab.binding.airquality/pom.xml:
```diff
@@ -0,0 +1,19 @@
+  </parent>
+
+  <groupId>org.openhab.binding</groupId>
+  <artifactId>org.openhab.binding.airquality</artifactId>
```
Please remove groupId and version so they would be inherited from the "parent" section.

hakan42 commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,303 @@
+            } else {
+                resultOk = true;
+
+                if (result.getData() == null) {
```
Maybe change this to "result.getData() != null" and only then set resultOk to true? the if clause is slightly confusing 😄

hakan42 commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,303 @@
+        // Update the thing status
+        if (resultOk && result != null) {
+            String attributions = result.getData().getAttributions();
+            logger.info("Received AirQuality data from aqicn.org. {}", attributions);
```
Do we need this result attribution (at info level) at each and every data query? Maybe it is good enough to show it only on Thing initialization?

hakan42 commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,303 @@
+     * Iterates through the fields and returns the getter value.
+     */
+    @SuppressWarnings("all")
+    private static Object getValue(Object data, String[] fields, int index) throws Exception {
```
Is reflection really necessary to fetch attributes from the json object? You might run into security manager problems...

hakan42 commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/internal/AirQualityConfiguration.java:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @author Kuba Wolanin - Initial contribution
+ */
+public class AirQualityConfiguration {
```
Use private members and public getter methods on the Configuration object please. This construct is slightly surprising for other Java programmers 😈

kubawolanin replied:
You got it! :) Thank you

kaikreuzer replied:
No, the idea of the [Configuration classes](https://www.eclipse.org/smarthome/documentation/development/bindings/how-to.html#configuration) is actually to have public fields and no need for boilerplate methods. This is also done in [other bindings](https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.autelis/src/main/java/org/openhab/binding/autelis/config/AutelisConfiguration.java) that way.

kubawolanin replied:
Reverted the chnge here so we have only public fields in Configuration class.

hakan42 commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/internal/AirQualityJsonResponse.java:
```diff
@@ -0,0 +1,108 @@
+    public String getAqiDescription() {
+        int aqi = data.getAqi();
+
+        if (aqi > 0 && aqi <= 50) {
```
Could you introduce String constants for those values?

hakan42 commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/internal/AirQualityJsonTime.java:
```diff
@@ -0,0 +1,33 @@
+
+    public Calendar getS() throws Exception {
+        Calendar calendar = Calendar.getInstance();
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
Please create a static member. Creating the SimpleDateFormat every time might be costly.

hakan42 commented on addons/binding/pom.xml:
```diff
@@ -71,4 +72,4 @@
   </modules>
 
-</project>
+</project>
```
Please do not remove the empty line 😄

hakan42 commented on addons/binding/org.openhab.binding.airquality/pom.xml:
```diff
@@ -0,0 +1,19 @@
+  <parent>
+    <groupId>org.openhab.binding</groupId>
+    <artifactId>pom</artifactId>
+    <version>2.0.0-SNAPSHOT</version>
```
Oh, by the way, we are at 2.1.0-SNAPSHOT as of four hours ago:

https://github.com/openhab/openhab2-addons/commit/af6fe5b1be64a009693157e733e2cd36952e00ae

I guess I need to update my pull requests as well 😄

ThomDietrich commented on addons/binding/org.openhab.binding.airquality/README.md:
```diff
@@ -0,0 +1,118 @@
+
+| Channel ID | Item Type    | Description              |   |
+|------------|--------------|------------------------- |---|
+| aqiLevel | Number | Air Quality Index | |
```
Is the last column reserved for something or a mistake?

kubawolanin replied:
It was certainly a mistake :) Thanks Thomas!

ThomDietrich commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,303 @@
+        super.initialize();
+
+        AirQualityConfiguration config = getConfigAs(AirQualityConfiguration.class);
+        logger.debug("config apikey = {}", config.apikey);
```
Should this be omitted from log files?

rtvb commented:
Thanks for your hard work.

These are some of my initial findings:

- README: Example things file uses `airquality:aqi:Krakow` but example items use `airquality:aqi:krk`. This is inconsistent. I would use `home` for both anyway for the example.
- README: You could move the .map example to the full example chapter.
- README: The .map file is missing an entry for value `-`. This raises an error in the console when used in the UI. I added an additional line `-=-` to fix this.
- README: The example things uses `location="50.06465,19.94498"` but the documentation says the `37.8;-122.4` syntax is allowed. This is confusing or even wrong?
- README: The example .items could use specification of icons
- README: You could prefix any item in the example with `Aqi_` for better readability
- README: It would be awesome if the text color of the AQI level and description matches the index. This could be achieved in a sitemap using visibility.

I would gladly help you get the docs right, but am not into git too much to know what is the best way to clone your work and add to the PR.

rtvb commented on addons/binding/org.openhab.binding.airquality/README.md:
```diff
@@ -0,0 +1,118 @@
+
+`AQI Description` item provides a human-readable output that can be interpreted e.g. by MAP transformation:
+ 
+airquality.map
```
You could move the .map example to the full example section

rtvb commented on addons/binding/org.openhab.binding.airquality/README.md:
````diff
@@ -0,0 +1,118 @@
+UNHEALTHY=Unhealthy
+VERY_UNHEALTHY=Very unhealthy
+HAZARDOUS=Hazardous
+```
````
The mapping `-=-` is missing. This raises an error when needed in a sitemap.

kubawolanin replied:
Done. Thanks!

rtvb commented on addons/binding/org.openhab.binding.airquality/README.md:
```diff
@@ -0,0 +1,118 @@
+Number AqiLevel "Air Quality Index" (AirQuality) { channel="airquality:aqi:krk:aqiLevel" }
+String AqiDescription "AQI Level [MAP(airquality.map):%s]" (AirQuality) { channel="airquality:aqi:krk:aqiDescription" }
+
+Number Pm25 "PM25 Level" (AirQuality) { channel="airquality:aqi:krk:pm25" }
```
These numbers could be enhanced with `<line>` icon.

rtvb commented on addons/binding/org.openhab.binding.airquality/README.md:
````diff
@@ -0,0 +1,118 @@
+```
+airquality:aqi:Krakow "AirQuality in Krakow" [ apikey="XXXXXXXXXXXX", location="50.06465,19.94498", refresh=60 ]
+```
+
````
Suggesting to prepend all items with 'Aqi_'

kubawolanin commented:
@rtvb @ThomDietrich @hakan42 @splatch Gentlemen, thank you so much for your tremendous help!
I think I've applied all of the changes needed. 

I've updated my first post with compiled JAR file.

Thanks again! 😃

ThomDietrich commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,308 @@
+
+        AirQualityConfiguration config = getConfigAs(AirQualityConfiguration.class);
+        logger.debug("config location = {}", config.getLocation());
+        logger.debug("config refresh = {}", config.getRefresh());
```
Hey! You've removed the line, which is totally fine. What I actually meant to suggest was to output a hardcoded "config apikey = (omitted from logging)". This way the user still knows/has the feeling his configuration was taken into account. I believe that is the actual purpose of these logger lines!?
This is a mere unimportant detail. Do as you feel it's right ;)

kubawolanin replied:
Done! Thank you:)

rtvb commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,308 @@
+            case AirQualityBindingConstants.LOCATIONURL:
+                return data.getData().getCity().getUrl();
+            case AirQualityBindingConstants.OBSERVATIONTIME:
+                return data.getData().getTime().getS();
```
To respect time zones you should probably parse the 'v' tag (Epoch) or add 'tz'

rtvb replied:
Specifically for the purpose that @kaikreuzer explicitly asked for reviewers' approval, I've started a new review with required changes. The timezone bug should be finished before merging.

kubawolanin replied:
Done!

kubawolanin replied:
Done! Thanks for review!

rtvb commented on addons/binding/org.openhab.binding.airquality/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,164 @@
+        </config-description>
+    </thing-type>
+
+    <channel-type id="aqiLevel">
```
Channel id matches actual value better if it is called `aqi` or `aqiId`

rtvb commented on addons/binding/org.openhab.binding.airquality/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,164 @@
+        <state readOnly="true" pattern="%d"></state>
+    </channel-type>
+
+    <channel-type id="aqiDescription">
```
According to the [website](http://aqicn.org/scale/), this value is referred to as "scale"  or "level". Channel id could perhaps better match this, so instead of `aqiDescription` use `aqiLevel`. Please note: you are using level for the actual index.
In my opinion an index is a value where a level is a range of values.

kubawolanin replied:
I don't think it makes much of a difference. I've had `aqiLevel` named as `aqi` but this conflicted with the thing type I wanted to keep short.
Let's leave it like it is now, I think it's descriptive.

rtvb replied:
Your use of the word level is different than that from the website. This might lead to confusion. But hey, it's your binding, so do as you feel is best.

rtvb commented on addons/binding/org.openhab.binding.airquality/README.md:
```diff
@@ -0,0 +1,128 @@
+
+Number   Aqi_Pm25            "PM25 Level" <line> (AirQuality) { channel="airquality:aqi:home:pm25" }
+Number   Aqi_Pm10            "PM10 Level" <line> (AirQuality) { channel="airquality:aqi:home:pm10" }
+Number   Aqi_O3              "O3 Level" <line> (AirQuality) { channel="airquality:aqi:home:o3" }
```
Replace label with "O\u2083 Level" to show subscript 3

rtvb commented on addons/binding/org.openhab.binding.airquality/README.md:
```diff
@@ -0,0 +1,128 @@
+Number   Aqi_Pm25            "PM25 Level" <line> (AirQuality) { channel="airquality:aqi:home:pm25" }
+Number   Aqi_Pm10            "PM10 Level" <line> (AirQuality) { channel="airquality:aqi:home:pm10" }
+Number   Aqi_O3              "O3 Level" <line> (AirQuality) { channel="airquality:aqi:home:o3" }
+Number   Aqi_No2             "NO2 Level" <line> (AirQuality) { channel="airquality:aqi:home:no2" }
```
Replace label with "NO\u2082 Level" to show subscript 2

kubawolanin replied:
Sure, why not!:)

rtvb replied:
Same goes for Pm25 and pm10. Too bad I haven't been able to locate a subscript dot (for subscript 2.5).

ThomDietrich commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/internal/json/AirQualityJsonTime.java:
```diff
@@ -0,0 +1,52 @@
+     * @return {String}
+     */
+    public String getTz() {
+        return tz.replace(":", "");
```
I'm looking at this on GitHub only, so sorry if I missed that: Where it `tz` initialized?

kubawolanin replied:
Right [there](https://github.com/openhab/openhab2-addons/pull/1738/files/a1559825c1407b4775865f52ba4117881bd0e4d9#diff-4287385b36535d4d39b5c779ff875befR25) :)

ThomDietrich replied:
😃 okay, let me more precise: where is it set?

kubawolanin replied:
It's being transformed by Gson from the response to a property.

ThomDietrich replied:
Ah, I see it now.
To be honest, I have little experience with serialization. Is there a better way to make that clear in source code? One has to go three classes "up" to even find the gson conversion.

kubawolanin replied:
Well, It's a good practice for classes [to be responsible for one thing](https://en.wikipedia.org/wiki/Single_responsibility_principle). 
In this case @splatch moved all JSON parsing-related logic to `org.openhab.airquality.intrernal.json` package. 
In `Handler` itself, we're executing Gson so it knows where to put all data
This gives ability to scale the classes in case of e.g. API change or new entry points coming with the payload :)
Cheers

splatch replied:
I was checking that and there is no nice way to convert time representation given by aqicn to `java.util.Date` or `Calendar` directly. Sadly there is none. With given data
`{v: int, s: string, tz:"01:00"}` we could convert unix timestamp coded in `v` field of json to java timestamp (coded in millis), but then we have to do dance with timezone. This means that only custom Gson `TypeAdapter` will cover that. In this way we would close time related handling in separate type, but since there is no other uses of this structure benefits are minor.  Without joda time or java 8 date-time api it will always be counter intuitive.
One thing we could do is change of simple date fromat to `yyyy-MM-dd HH:mm:ssX` to parse concatenated `s+ tz` fields easily. `X` allows to specify time zone offset in format returned by service.

rtvb replied:
Taking into account my note below, I think getting the proper time is to simple concatenate the `s` and `tz` values and parse them with `DateTimeFormatter` with `ISO_OFFSET_DATE_TIME` or something similar?

https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_OFFSET_DATE_TIME

rtvb commented:
I did some additional testing with time zones. And notice some differences for Rotterdam and Moscow. Perhaps the date in the JSON is local time and does not need modification after all. See issue:

https://github.com/kubawolanin/openhab2-addons/issues/2

ThomDietrich commented:
@kubawolanin my "where is tz set" question was still bugging me.
Here is why: Besides the small comment above the class definition, there is no clear hint to any special nature of that class. All I was seeing was a null-initialized String and a replace call - which last time I checked will throw an exception. The actual meaning got clear to me only after you pointed it out, way outside of the class scope.

The class should be annotated in someway to make the gson relation clear. It seems, [`SerializedName`](https://google.github.io/gson/apidocs/com/google/gson/annotations/SerializedName.html) would be an approach here. @splatch are there better approaches to make the software design clear and the class self-contained and sane?

kubawolanin commented:
@ThomDietrich I've added some more annotations and `SerializedName` for the values as well. Hope that'll do :)

@rtvb thank you for this. I'm no longer modifying it.

BTW. I've updated my first post with latest build.
Per [discussion on the forum](https://community.openhab.org/t/air-quality-binding-looking-for-testers/20756/30?u=kubawolanin) - there was an update in the API that caused trouble in displaying PM2.5, PM10 etc. 

The latest build also includes:

* More descriptive error display (e.g. if there's something wrong with config)
* Fixed docs + added Unicode chars
* Removed Station ID from the channels (it's no longer there in the payload)
* Better Location description (I've deleted `"` from the example to avoid confusion)
* Changed measurements from int to floats (per API change)

Thank you all for the cooperation!

rtvb commented:
I will do some additional testing for NYC, London and Moscow re. time zones.

rtvb commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/internal/json/AirQualityJsonTime.java:
```diff
@@ -0,0 +1,56 @@
+    @SerializedName("timeZone")
+    private String tz;
+
+    private final static SimpleDateFormat SDF = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
Why not use "yyyy-MM-dd'T'HH:mm:ssZ" as the pattern and add the `tz` to the string when creating the date?

rtvb commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/internal/json/AirQualityJsonTime.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    public Calendar getDateString() throws Exception {
+        Calendar calendar = Calendar.getInstance();
+        Date date = SDF.parse(dateString);
```
Perhaps replace with:
```
Date date = SDF.parse(dateString + tz);
```
If adding the timezone to the formatter and the build string, shouldn't you get the observation time in the correct time zone?

kubawolanin commented:
@rtvb then I misunderstood [your post](https://github.com/openhab/openhab2-addons/pull/1738#issuecomment-274604956).
Here's your whole setup on my environment - I've applied date+timezone concatenation again and used `yyyy-MM-dd'T'HH:mm:ssZ` format.
Here's the result. Would that satisfy your requrement?
![clipboard02](https://cloud.githubusercontent.com/assets/2270505/22403778/2aa5461c-e622-11e6-8153-1c8cfb50ca92.jpg)

rtvb commented:
All times now seem to be in my local zone. That would satisfy my needs

kubawolanin commented:
Hey @rtvb - I've pushed my latest changes to the repo!
Feel free to test it - I hope it's ok now! 😃

rtvb commented:
Cool updates to units you did in that commit https://github.com/openhab/openhab2-addons/pull/1738/commits/612a21d000faf63938d05d88f73ae39f72ce2112

I will try the new binding the next coming days and report back.

hakan42 commented:
https://airvisual.com/world seems to use the same data, the stations in my area of the world are exactly the same as shown by http://aqicn.org/ ... Maybe something to add to the README.md as "Conrad Connect" in Germany advertises that they have a channel for airvisual.com now?

kubawolanin commented:
> Maybe something to add to the README.md as "Conrad Connect" in Germany advertises that they have a channel for airvisual.com now?

You propose adding info about alternative API to the README? I don't think I understand.
Maybe in the future we could choose (in config?) where you want to get the data from, but I think it won't not useful for now.

hakan42 commented:
No, I was not clear on my message. Both airvisual and aqicn seem to use the same data source, so the implementation is o.k.

I was just thinking about advertising in the README.md that we can already do what Conrad just started to implement 😄

kubawolanin commented:
I've updated my first post with the latest build.
What's new:
* I've added [a simple rule example](https://github.com/kubawolanin/openhab2-addons/tree/feature/airquality/addons/binding/org.openhab.binding.airquality#full-example) at the bottom of `README.md`
* Added a possibility to [query data from single Station ID](https://github.com/kubawolanin/openhab2-addons/issues/10)
* Fixed [humidity formatting](https://github.com/kubawolanin/openhab2-addons/issues/9)
* [Warnings about NULL values](https://github.com/kubawolanin/openhab2-addons/issues/8) are logged in DEBUG mode now
* `location` parameter is not `required` now in the PaperUI, since you can specify `stationId` instead. 

Thank you!

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,177 @@
+		<state readOnly="true" pattern="%s"></state>
+	</channel-type>
+
+	<channel-type id="locationUrl" advanced="true">
```
Is this channel really useful? We do not support links in item states, so I do not know, how this should be used in any meaningful way?

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/README.md:
```diff
@@ -0,0 +1,197 @@
+| stationId | Unique ID of the measuring station. |
+| refresh   | Refresh interval in minutes. Optional, the default value is 60 minutes.  |
+
+For the location parameter, the following syntaxes are allowed (comma or semicolon):
```
see comment above - imho comma is enough

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,177 @@
+	</channel-type>
+
+	<channel-type id="stationLocation" advanced="true">
+		<item-type>String</item-type>
```
This should clearly be a "Location" item, shouldn't it?

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/about.html:
```diff
@@ -0,0 +1,47 @@
+<p>January 20, 2017</p>	
+<h3>License</h3>
+
+<p>The Eclipse Foundation makes available all content in this plug-in (&quot;Content&quot;).  Unless otherwise 
```
This somehow got lost in the docs. It isn't the Eclipse Foundation, but the openHAB Community that makes this available - see [here](https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.avmfritz/about.html) for a correct version of the about file.

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,321 @@
+ * sent to one of the channels.
+ *
+ * @author Kuba Wolanin - Initial contribution
+ * @author Łukasz Dywicki
```
make sure to mention him with an "Also-By" line in your commit message (and also make sure that he is ok with the contribution and the DCO, but I guess there is no doubt in that case).

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,321 @@
+        boolean validConfig = true;
+
+        if (StringUtils.trimToNull(config.getApiKey()) == null) {
+            logger.error("Parameter 'apikey' is mandatory and must be configured, disabling thing '{}'",
```
Don't do error logging, but simply provide this info to the user by setting the Thing status accordingly (third parameter in the updateStatus call).

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,321 @@
+            validConfig = false;
+        }
+        if (StringUtils.trimToNull(config.getLocation()) == null && config.getStationId() == null) {
+            logger.error(
```
dito

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,321 @@
+            validConfig = false;
+        }
+        if (config.getRefresh() != null && config.getRefresh() < 5) {
+            logger.error("Parameter 'refresh' must be at least 5 minutes, disabling thing '{}'", getThing().getUID());
```
dito

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,321 @@
+                            updateChannel(channel.getUID().getId(), aqiResponse);
+                        }
+                    } catch (Exception e) {
+                        logger.error("Exception occurred during execution: {}", e.getMessage());
```
Check against [the guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging): Is it a bug if this code is reached? Then also provide the `e` as a last parameter, so that the stack trace is included. Or can this also e.g. happen if the HTTP request times out? Then rather set the Thing status instead.

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,321 @@
+
+            AirQualityConfiguration config = getConfigAs(AirQualityConfiguration.class);
+            int period = (config.getRefresh() != null) ? config.getRefresh().intValue() : DEFAULT_REFRESH_PERIOD;
+            refreshJob = scheduler.scheduleAtFixedRate(runnable, 0, period, TimeUnit.MINUTES);
```
See D.2 in the guidelines - use scheduleWithFixedDelay instead

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,321 @@
+        if (command instanceof RefreshType) {
+            updateChannel(channelUID.getId(), aqiResponse);
+        } else {
+            logger.warn("The Air Quality binding is a read-only and can not handle command {}", command);
```
typo: "is read-only"

kaikreuzer replied:
Imho, this should only be debug. Maybe the channel is linked to an item, which is also linked to other bindings, which accept commands. So sending a command to the item might be fully correct - your binding should simply ignore it.

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,321 @@
+    /**
+     * Get new data from aqicn.org service
+     *
+     * @return {boolean}
```
boolean?

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,321 @@
+            // Run the HTTP request and get the JSON response from aqicn.org
+            URL url = new URL(urlStr);
+            URLConnection connection = url.openConnection();
+            String repsonse = IOUtils.toString(connection.getInputStream());
```
typo: response

kaikreuzer replied:
Better do a try-with-resources as [IOUtls.toString does not close the input stream by itself](https://commons.apache.org/proper/commons-io/description.html).

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java:
```diff
@@ -0,0 +1,321 @@
+            logger.warn("Error running aqicn.org (Air Quality) request: {}", errorMsg);
+        } catch (IOException | IllegalStateException e) {
+            errorMsg = e.getMessage();
+            logger.warn("Error running aqicn.org (Air Quality) request: {}", errorMsg);
```
I would expect IOExceptions to happen if you do not have a network connection, right?
So according to the guidelines, you must not log a warning here, but set the Thing status accordingly only (what you do below).

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/internal/json/AirQualityJsonCity.java:
```diff
@@ -0,0 +1,46 @@
+        for (int i = 0; i < geo.size(); i++) {
+            list.add(geo.get(i).toString());
+        }
+        return String.join(", ", list);
```
join only exists in Java8, while your MANIFEST.MF states Java7 as "Bundle-RequiredExecutionEnvironment". You should adapt the MANIFEST.MF.

kaikreuzer commented:
> @kaikreuzer the PR gained two approvals so pinging you for awareness ;-)

@rtvb & @ThomDietrich: Thanks for helping on the review! Let us try to align about the requirements that need to be met before approving a PR - the main checklist should be the [guidelines](http://docs.openhab.org/developers/development/guidelines.html). If you go through my comment, you will notice that a couple of things are still missing there; I'll try to check where to best put such infos (like e.g. about .project & feature.xml files).

kubawolanin commented:
@kaikreuzer thank you for the review! I'll address the comments as soon as I find some time for it (hopefully this week).

kubawolanin commented:
@kaikreuzer I've applied all of the comments. Thanks again for your insight!

Regarding the binding itself, a few ***breaking changes***:
1. `Location` configuration no longer understands semicolon - you need to use comma instead (which is now compliant with other bindings)
1. I've removed `stationUrl` channel - I couldn't find any reason to keep it. Let me know if you find any use case for it!
1. `Aqi_StationGeo` item (and `stationLocation` channel) is now a `Location` element (instead of `String`) which allows you to display it on a map (closes kubawolanin#1)

I will update my first post (and the [community thread](https://community.openhab.org/t/air-quality-binding-looking-for-testers/20756)) with latest compiled JAR in a few hours from now.

Thanks again! :sunglasses:

kaikreuzer commented on addons/binding/org.openhab.binding.airquality/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,27 @@
+Bundle-Name: AirQuality Binding
+Bundle-SymbolicName: org.openhab.binding.airquality;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.0.0.qualifier
```
One very last issue: This needs to be 2.1.0.qualifier!

kubawolanin replied:
Forgot about this one. Thanks! :)

kaikreuzer commented:
perfect, you are quick! So no need to share another jar - I'll directly run a new official build, which will contain it 👍

rtvb commented:
Will the README be added to the docs site automatically or do we need to create an additional issue or pull request?

kaikreuzer commented:
Already there as you can see: http://docs.openhab.org/addons/bindings/airquality/readme.html

emme-g commented:
Hello,
some days ago I started to use this useful binding and I could set almost everything in an easy and quick way.
In the last days I was trying to understand in a better way how the AQI is calculated because some values I could read through the binding were a little unclear for me.
I understood that the total AQI is calculated as the maximum value among the individual AQIs related to the single pollutants (PM2,5, PM10, O3, NO2, CO), as it is better explained in Wikipedia
https://en.wikipedia.org/wiki/Air_quality_index.
I noticed that in the air quality binding the values of the items related to the single pollutants contain the μg/m³ dimension, but it seems to me that the individual AQUIs related to the single pollutants are indexes and not values with μg/m³ dimensions, as we can read for example in this page
http://aqicn.org/faq/2013-09-09/revised-pm25-aqi-breakpoints/
where it is explained that a value 50 of the PM2,5 individual AQI corresponds to a PM2,5 concentration of 12 μg/m³.
So, I think that it should be better that the item values would appear in the binding without any dimension.
It would be very interesting to receive the values of the concentrations of the single pollutants too, but I think that this depends on the AQIcn.org service and eventually it should be ask to them.
Thank you for your work, I hope this opinion could be help someway.

kevinshane commented:
Hi, I am using docker openhab 2.1.0 stable version with airquality binding installed, howerver I have some strange errors: when restarting the OS, the airquality binding go online at startup and I get all the weather info, but the binding **immediately** changed from online back to initializing and then go through the process again, finally go online once again, but this time I lose all the weather info, every item changed from current value to UNDEF.. and never go back.. what's the problem here? 
here is the log:
2017-12-01 20:57:50.875 [ItemStateChangedEvent     ] - Aqi_Level changed from NULL to 68
2017-12-01 20:57:50.878 [ItemStateChangedEvent     ] - Aqi_Description changed from NULL to MODERATE
2017-12-01 20:57:50.882 [ItemStateChangedEvent     ] - Aqi_Pm25 changed from NULL to 68
2017-12-01 20:57:50.887 [ItemStateChangedEvent     ] - Aqi_Pm10 changed from NULL to 45
2017-12-01 20:57:50.889 [ItemStateChangedEvent     ] - Aqi_O3 changed from NULL to 53.2
2017-12-01 20:57:50.892 [ItemStateChangedEvent     ] - Aqi_No2 changed from NULL to 10.6
2017-12-01 20:57:50.895 [ItemStateChangedEvent     ] - Aqi_Co changed from NULL to 9.1
2017-12-01 20:57:50.899 [ItemStateChangedEvent     ] - Aqi_Temperature changed from NULL to 21.15
2017-12-01 20:57:50.909 [ItemStateChangedEvent     ] - Aqi_Pressure changed from NULL to 1019
2017-12-01 20:57:50.912 [ItemStateChangedEvent     ] - Aqi_Humidity changed from NULL to 68
2017-12-01 20:57:51.034 [hingStatusInfoChangedEvent] - 'airquality:aqi:home' changed from UNINITIALIZED to INITIALIZING
2017-12-01 20:57:51.087 [hingStatusInfoChangedEvent] - 'airquality:aqi:home' changed from INITIALIZING to ONLINE
2017-12-01 20:57:51.091 [ItemStateChangedEvent     ] - Aqi_Level changed from 68 to UNDEF
2017-12-01 20:57:51.095 [ItemStateChangedEvent     ] - Aqi_Description changed from MODERATE to UNDEF
2017-12-01 20:57:51.119 [ItemStateChangedEvent     ] - Aqi_Pm25 changed from 68 to UNDEF
2017-12-01 20:57:51.121 [ItemStateChangedEvent     ] - Aqi_Pm10 changed from 45 to UNDEF
2017-12-01 20:57:51.124 [ItemStateChangedEvent     ] - Aqi_No2 changed from 10.6 to UNDEF
2017-12-01 20:57:51.139 [ItemStateChangedEvent     ] - Aqi_O3 changed from 53.2 to UNDEF
2017-12-01 20:57:51.142 [ItemStateChangedEvent     ] - Aqi_Co changed from 9.1 to UNDEF
2017-12-01 20:57:51.144 [ItemStateChangedEvent     ] - Aqi_Temperature changed from 21.15 to UNDEF
2017-12-01 20:57:51.146 [ItemStateChangedEvent     ] - Aqi_Pressure changed from 1019 to UNDEF
2017-12-01 20:57:51.149 [ItemStateChangedEvent     ] - Aqi_Humidity changed from 68 to UNDEF
2017-12-01 20:57:52.129 [ERROR] [airquality.handler.AirQualityHandler] - Exception occurred during execution: null
java.lang.NullPointerException
	at org.openhab.binding.airquality.internal.json.AirQualityJsonData.getAttributions(AirQualityJsonData.java:86)[194:org.openhab.binding.airquality:2.1.0]
	at org.openhab.binding.airquality.handler.AirQualityHandler.getAirQualityData(AirQualityHandler.java:270)[194:org.openhab.binding.airquality:2.1.0]
	at org.openhab.binding.airquality.handler.AirQualityHandler.updateAirQualityData(AirQualityHandler.java:198)[194:org.openhab.binding.airquality:2.1.0]
	at org.openhab.binding.airquality.handler.AirQualityHandler.access$0(AirQualityHandler.java:196)[194:org.openhab.binding.airquality:2.1.0]
	at org.openhab.binding.airquality.handler.AirQualityHandler$1.run(AirQualityHandler.java:114)[194:org.openhab.binding.airquality:2.1.0]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_144]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)[:1.8.0_144]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)[:1.8.0_144]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)[:1.8.0_144]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)[:1.8.0_144]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)[:1.8.0_144]
	at java.lang.Thread.run(Thread.java:748)[:1.8.0_144]

kubawolanin commented:
@emme-g, @kevinshane - please submit separate issues for your requests. Thanks!
